internal/api/v1/services.go

```diff
@@ -0,0 +1,157 @@
+package v1
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/epinio/epinio/helpers/kubernetes"
+	"github.com/epinio/epinio/internal/application"
+	"github.com/epinio/epinio/internal/cli/clients"
+	"github.com/epinio/epinio/internal/services"
+	"github.com/julienschmidt/httprouter"
+)
+
+type ServicesController struct {
+}
+
+func (sc ServicesController) Show(w http.ResponseWriter, r *http.Request) {
+	params := httprouter.ParamsFromContext(r.Context())
+	org := params.ByName("org")
+	serviceName := params.ByName("service")
+
+	cluster, err := kubernetes.GetCluster()
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+	gitea, err := clients.GetGiteaClient()
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+
+	exists, err := gitea.OrgExists(org)
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+	if !exists {
+		http.Error(w, fmt.Sprintf("Organization '%s' does not exist", org),
+			http.StatusNotFound)
+		return
+	}
+
+	service, err := services.Lookup(cluster, org, serviceName)
+	if err != nil {
+		if err.Error() == "service not found" {
+			http.Error(w, fmt.Sprintf("Service '%s' does not exist", serviceName),
+				http.StatusNotFound)
+			return
+		}
+		if handleError(w, err, http.StatusInternalServerError) {
+			return
+		}
+	}
+
+	status, err := service.Status()
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+	serviceDetails, err := service.Details()
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+
+	responseData := map[string]string{
+		"Status": status,
+	}
+	for key, value := range serviceDetails {
+		responseData[key] = value
+	}
+
+	js, err := json.Marshal(responseData)
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+	w.Header().Set("Content-Type", "application/json")
+	_, err = w.Write(js)
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+}
+
+func (sc ServicesController) Index(w http.ResponseWriter, r *http.Request) {
+	params := httprouter.ParamsFromContext(r.Context())
+	org := params.ByName("org")
+
+	cluster, err := kubernetes.GetCluster()
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+
+	gitea, err := clients.GetGiteaClient()
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+	exists, err := gitea.OrgExists(org)
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+	if !exists {
+		http.Error(w, fmt.Sprintf("Organization '%s' does not exist", org),
+			http.StatusNotFound)
+		return
+	}
+
+	orgServices, err := services.List(cluster, org)
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+	appsOf, err := servicesToApps(cluster, gitea, org)
+	if handleError(w, err, http.StatusInternalServerError) {
+		return
+	}
+
+	responseData := map[string]interface{}{
+		"Services":    orgServices,
+		"ServiceApps": appsOf,
```

Comment:
Interesting ... You are returning the list of services, and the bound apps as separate keys of an object ...

I would have used an array of objects, with the info about bounds app added to the service objects directly.

Hm. Another point here. While the marshalling and transmission is easy, trivial even, it also transmits everything the server knows, even stuff we may consider internal (Instance name for catalog services, and secret name for the custom services).

Reply:
And looking at the client side, with all the casting, that makes me nervous.
I am coming to the opinion that a custom struct type should be used for the information, instead of the internal service structures. Something like

```
type struct ServiceListReponse {
    Name: string
    BoundApp: []string
}
```

While using that would require the server to convert from the internal form to this, the cli side unmarshalling becomes much simpler/easier, and no casting will be required at all for get the information to print.

Reply:
A custom response struct would also fit nicely with generated API docs (The wrappers can reference those structs https://github.com/epinio/epinio/pull/272/files#diff-e36e12b03db99680c15305336f9147bfac43be780989ad2833633a457839f4c8R36)

Reply:
I like the special struct idea. Adheres to the Presenter pattern and it separates the view logic.

Reply:
Done.